src/plugins/intel_gpu/src/plugin/sync_infer_request.cpp

```diff
@@ -444,13 +444,12 @@ void SyncInferRequest::wait() {
             auto& stream = m_graph->get_network()->get_stream();
             auto user_mem = remote_tensor_impl_ptr->get_original_memory();
             if (user_mem->get_allocation_type() == cldnn::allocation_type::cl_mem && output_memory->get_allocation_type() != cldnn::allocation_type::cl_mem) {
-                // WA: Copy between cl_mem and usm memory may fail for some reason (driver bug?)
-                // so this explicit memcpy is used to provide correct output for cl_mem output in dynamic cases
-                cldnn::mem_lock<uint8_t, cldnn::mem_lock_type::write> lock_dst(user_mem, stream);
-                cldnn::mem_lock<uint8_t, cldnn::mem_lock_type::read> lock_src(output_memory, stream);
-                std::memcpy(lock_dst.data(), lock_src.data(), output_memory->size());
+                auto blocking = (stream.get_queue_type() == QueueTypes::in_order) ? false : true;
+                copy_events.push_back(user_mem->copy_from(stream, *output_memory, blocking));
             } else {
-                copy_events.push_back(output_memory->copy_to(stream, *user_mem, false));
+                bool is_same_mem = output_memory->buffer_ptr() == user_mem->buffer_ptr();
```

Comment:
1) These two branches are almost identical now, probably they can be combined?
2) Engine's `is_the_same_buffer()` can be used here
3) As we save copy event, do we need to make call blocking for out_of_order queue?

Reply:
merged two branches into one as you suggested